### PR TITLE
Code 148

### PR DESCRIPTION
**Pull request by ArBridgeman**
- [x] Merge code-154
- [x]  Merge code-157
- [x] rebase off master
- [x] check code & all menus that now have (weekday) split to (weekday, meal_time), (factor, unit)->(eat_factor, eat_unit), and a new column freeze_factor.
- [x] squash all commits together (note that autosquash would not properly handle)
- [x] merge CODE-148

TODOS
- [x]  finish implementing main menu tests
- [x] add way to handle freeze_factor to todoist
- [x] fix items that use duedate & todoist
- [x]  in grocery list add freeze_factor & eat_factor for recipes
- [x] create unit tests for duedate
- [x] create unit tests for todoist
- [x] create unit tests for grocery list
- [x] in grocery list, only have weekday of menu, not meals
- [x] add in menu way to handle times for different meals & test

### CODE REVIEW

File: sous_chef/sous_chef/date/get_due_date.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+
+class MealTime(Enum):
+    breakfast = {"hour": 8, "minute": 30}
+    lunch = {"hour": 11, "minute": 30}
+    dinner = {"hour": 18, "minute": 15}
+
+
+class DueDatetimeFormatter:
+    # TODO add anchor date, default time in config?
```
Comment (AkagiShigeru) [inline sous_chef/sous_chef/date/get_due_date.py:23]: Yes, would do later, not super critical.

**Review by AkagiShigeru — APPROVED** [review]
Thanks for the wealth of additions and useful tests. I had a few suggestions that I committed, nothing major I feel. Feel free to have a look. All tests are working for me locally.

Some other comments and remarks:
Regarding `test_todoist_api.py`
- As most of the untested functions use the todoist api, I don't think these need to be explicitly tested.
- Some of the code could probably be refactored to make more elements testable but I wouldn't give that a high importance right now.

Regarding `test_generate_grocery_list.py`:
- That was a huge amount of work, wow
- I added code for the test `test__add_referenced_recipe_to_queue`, probably not in a very nice way, please check
- Of the two untested remaining functions, it might make sense to test `_aggregate_group_to_grocery_list`, which should be relatively simple, but not very important, I'd say. I'm fine if this is not done, there are more interesting things in the pipeline ;).
- `test__aggregate_grocery_list_by_item_and_dimension` completely consists of already tested functions, so should be fine not to implement this.

Otherwise, I think this was again a big step forward in test coverage, thanks a lot.

**Comment by ArBridgeman** [comment]
in _get_cooking_time_min, if the value is negative, why set it to 0? Wouldn't it be more consistent to set with default value of 20?

**Comment by ArBridgeman** [comment]
In test__add_referenced_recipe_to_queue, couldn't we perform 1 assert with a menu recipe in a list? Then we're not testing length and the attributes separately?

**Comment by AkagiShigeru** [comment]
> In test__add_referenced_recipe_to_queue, couldn't we perform 1 assert with a menu recipe in a list? Then we're not testing length and the attributes separately?

Agree, I think I know what you mean. I'll commit a suggested change for your consideration.

**Comment by AkagiShigeru** [comment]
Pushed minor changes to address two comments by @ArBridgeman.

**Comment by ArBridgeman** [comment]
> Thanks for the wealth of additions and useful tests. I had a few suggestions that I committed, nothing major I feel. Feel free to have a look. All tests are working for me locally.
> 
> Some other comments and remarks: Regarding `test_todoist_api.py`
> 
> * As most of the untested functions use the todoist api, I don't think these need to be explicitly tested.
> * Some of the code could probably be refactored to make more elements testable but I wouldn't give that a high importance right now.
> 
> Regarding `test_generate_grocery_list.py`:
> 
> * That was a huge amount of work, wow
> * I added code for the test `test__add_referenced_recipe_to_queue`, probably not in a very nice way, please check
> * Of the two untested remaining functions, it might make sense to test `_aggregate_group_to_grocery_list`, which should be relatively simple, but not very important, I'd say. I'm fine if this is not done, there are more interesting things in the pipeline ;).
> * `test__aggregate_grocery_list_by_item_and_dimension` completely consists of already tested functions, so should be fine not to implement this.
> 
> Otherwise, I think this was again a big step forward in test coverage, thanks a lot.

I decided not to add the grocery list test. While you had a good point, I wanted to focus my work today & tomorrow on adding the new feature (random by tag) so that I do not need to select them myself this week.

**Pull request merged**